bookdown learnings (#624)
ropensci/roweb2: content/blog/2020-03-01-bookdown-learnings/index.Rmd

---
slug: "bookdown-learnings"
title: 10 Things We Learned in Creating the Blog Guide in bookdown
authors:
  - Stefanie Butland
  - Maëlle Salmon
date: 2020-03-01
tags:
  - bookdown
  - usethis
  - happygitwithr
  - github
  - rstudio
# the summary below will be used by e.g. Twitter cards
description: "10 bookdown tips for newbies and experts"
# If you have no preferred image for Twitter cards,
# delete the twitterImg line below 
# Note there is no '/' symbol before 'img' here
# if needed replace blog with technotes
#twitterImg: blog/2019/06/04/post-template/name-of-image.png
# 'output' is necessary to obtain index.md
# Do not commit index.html
output: 
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.path = "", comment = "", warning = FALSE)

knitr::knit_hooks$set(
  plot = function(x, options) {
    hugoopts <- options$hugoopts
    paste0(
      "{{<figure src=",
      '"', x, '" ',
      if (!is.null(hugoopts)) {
        glue::glue_collapse(
          glue::glue('{names(hugoopts)}="{hugoopts}"'),
          sep = " "
        )
      },
      ">}}\n"
    )
  }
)
```

> After soliciting, reviewing, and publishing over 100 blog posts and tech notes by rOpenSci community members, we have launched the [rOpenSci Blog Guidelines for Authors and Editors](https://blogguide.ropensci.org/). 

Author: ```suggestion
After soliciting, reviewing, and publishing over 100 blog posts and tech notes by rOpenSci community members, we have created the [rOpenSci Blog Guidelines for Authors and Editors](https://blogguide.ropensci.org/) to address many frequently asked questions and frequently given suggestions.. 
```

> We created it to address many frequently asked questions and frequently given suggestions.

Author: ```suggestion
```

Technically, we structured the content as a bookdown gitbook. It was Stef's first foray into the glorious process of publishing a book with bookdown, and Maëlle's second[^1]. And oh, we learned a lot.

### Stef (first-timer)

#### 1. How to collaborate across timezones

> There is a 9-hour time difference between Maëlle and me for most of the year. I'm in Kamloops, Canada and Maëlle is in Nancy, France. Since so much of this is new to me, text-based explanations from Maëlle via Slack usually boggled my mind. It helped immensely to have a weekly 30-minute  meeting (8:30am for Stef and 5:30pm for Maëlle) with agenda and notes in a shared google doc. We would talk through our approaches and priorities and Maëlle would coach me in new-to-me tools.

Reviewer: I don't think the short term time change is important enough to mention.

> There is a 9-hour time difference between Maëlle and me for most of the year. I'm in Kamloops, Canada and Maëlle is in Nancy, France. Since so much of this is new to me, text-based explanations from Maëlle via Slack usually boggled my mind. It helped immensely to have a weekly 30-minute  meeting (8:30am for Stef and 5:30pm for Maëlle) with agenda and notes in a shared google doc. We would talk through our approaches and priorities and Maëlle would coach me in new-to-me tools.

Reviewer: I do understand your point about saying "google" but in this case I feel like the tool choice is significant, just as GitHub is.

At the end of a day, I would DM Maëlle on Slack to say "I’m finished for the day, can you please review and merge my pull requests?" or "Please review the structure but not the text yet". Next morning, Maëlle would have done that plus her own work so I could update my local copy of our bookdown book and open new pull requests for new pieces of work.

Part way through the project, we got more strict about assigning issues to ourselves and tying them to specific milestones like "official release" or "nice to have one day". This helped us work asynchronously toward a common goal, showed some light at the end of the tunnel, as well as helping us (try to) avoid [scope-creep](https://en.wikipedia.org/wiki/Scope_creep).

#### 2. How to work in RStudio and GitHub to collaborate on a project

I'm an R beginner. 
My blog posts have never contained any R code. 

> I've always written them in Markdown using the Atom editor and I've always run git from the terminal, building on recipes that Scott Chamberlain and Karthik Ram have shared with me. 

Reviewer: I use Atom cause that's one of the free options that Scott mentioned. lol. not writing that.

After three years working in git and GitHub, I'm an intermediate user who still lacks robust mental model and the vocabulary to explain it[^2]. 
(This is not self-deprecation.) 
This bookdown Blog Guide project was the first time I used RStudio hooked up to GitHub to collaborate. 

> In the process, I have become much more comfortable with installing packages as needed and keeping R, RStudio, and package versions up to date with a weekly calendar reminder.

Reviewer: Do you mean what strategies to people use to force themselves to update their software? I decided on calendar after asking Scott what he does (updates every morning I think he said). The trick is, one has to actually do it when their calendar tells them to.

> In the process, I have become much more comfortable with installing packages as needed and keeping R, RStudio, and package versions up to date with a weekly calendar reminder.

Author: Such a good topic for the R-hub blog 🤔

- For setting up RStudio and knowing how to open a project from .Rproj I loved R-Ladies Sydney's "[opinionated tour of RStudio](https://rladiessydney.org/courses/ryouwithme/01-basicbasics-1/)". 
- I used happygitwithR "[Connect RStudio to Git and GitHub](https://happygitwithr.com/rstudio-git-github.html)" to do exactly that. 
Years ago I had already used [happygitwithR](https://happygitwithr.com/) to set up my git/GitHub from square-one and my https credentials.
- Maëlle and I both worked on feature branches of master. 

> On any given day we each might work on mulitple chapters of the book. Because each book chapter is generated from one Rmd file (more on that in tip #4), and you don't want one pull request to be _too_ complex, I would create a separate feature branch for any chapter or file I worked on. I would periodically push my work to GitHub in a pull request (more in tip #3) with "_Draft_" status so Maëlle could see I was not yet seeking her review. At the end of a day's commits, I would update the pull request status to "_Ready for review_" and assign Maëlle as reviewer. She would review, edit, and merge my pull requests and open new branches of her own for me to review.

Author: ```suggestion
On any given day we each might work on mulitple chapters of the book. Because each book chapter is generated from one Rmd file (more on that in tip #4), and you don't want one pull request to be _too_ complex, I would create a separate feature branch for any chapter or file I worked on. I would periodically push my work to GitHub in a pull request (more in tip #3) with ["_Draft_" status](https://github.blog/2019-02-14-introducing-draft-pull-requests/) so Maëlle could see I was not yet seeking her review. At the end of a day's commits, I would update the pull request status to "_Ready for review_" and assign Maëlle as reviewer. She would review, edit, and merge my pull requests and open new branches of her own for me to review.
```

#### 3. How to usethis

> At the start of the project, Maëlle asked me if I used the usethis package. 

Author: Dear native speaker, why is it "if" and not "whether"? (Why is English so complicated? :-) )

Reviewer: 🤔 I just make it up as I go 😉. I usually think of "whether" in "whether or not" and in this case there's no "not". It's entirely possible that my grammar isn't right here, and I'm comfortable with it in this case 😸

I did not. 

> But I bought into it because "[usethis](https://usethis.r-lib.org/) is a workflow package: it automates repetitive tasks that arise during project setup and development, both for R packages and non-package projects."

Author: ```suggestion
But I bought into it because _"[usethis](https://usethis.r-lib.org/) is a workflow package: it automates repetitive tasks that arise during project setup and development, both for R packages and non-package projects."_
```

Not because it's co-maintained by Jenny? :wink:

The trickiest part, for me, was commiting to setting up and forcing myself to use usethis pull request helpers despite not really understanding how it would help me. 

From previous work I had already followed happygitwithr to set up my [GitHub authentication via https](https://happygitwithr.com/credential-caching.html) and a [GitHub Personal Access Token (PAT)](https://happygitwithr.com/github-pat.html) in my `.Renviron`.

> I followed usethis [setup instructions](https://usethis.r-lib.org/articles/articles/usethis-setup.html) to edit my `.Rprofile` so that its functions would be available without explicitly calling the package.

Author: Fun fact (not to add in the post), I suggested this usethis article :sunglasses: https://www.tidyverse.org/blog/2018/02/usethis-1-3-0/#acknowledgements

Reviewer: coooool!
oh, and the real reason I bought into using usethis, is because you strongly suggested I should :-)

I also followed the rest of the setup instructions somewhat religiously (e.g. "Prepare your system to build packages from source"), because I wasn't confident enough to make a judgement on what I would and would not need.

Practically, I kept the usethis [pull request helpers](https://usethis.r-lib.org/articles/articles/pr-functions.html) article open in my browser and forced myself to use them.
At the start of the day, I would open our bookdown RStudio project by clicking my local `ropensci-blog-guidance.Rproj` file, pull to update my local master (because Maëlle would have reviewed, edited and merged my pull requests while I slept), create a new branch with `pr_init(branch = "branchname")`, make edits, and `pr_push()` to push my local changes to GitHub master.

> usethis automatically opened a browser to the GitHub web interface prompting me to open a pull request. 

Author: In the end, was it worth it? This tip lacks a conclusion. :-)

Reviewer: You caught me. Lack of conclusion was glaring to me because I think it was worth it, but it's tied up in my new use of RStudio so hard to separate the value of usethis alone.

Author: Maybe you could still say whether the PR workflow feels as good as what you do in the terminal?

#### 4. How to fill up a bunch of Rmd files to make a book

To answer the question, "how do I do this thing?" I often compared the 
GitHub file structure and contents of completed books with their corresponding live pages. This, for me, had the biggest payoff in learning how to bookdown.

> Things I learned:

Author: Please add empty lines between items :-)

- Create links inside the book with `{#anchor}`. In the [`authorcontent.Rmd`](https://github.com/ropensci-org/blog-guidance/blob/master/authorcontent.Rmd) file, the heading `# Content Guidelines {#content}` means that a markdown format `[link to that chapter](#content)` anywhere in the book will link to the Content Guidelines chapter.
- Appendices A to H (_too many; we know_) are created from a single [appendix.Rmd](https://github.com/ropensci-org/blog-guidance/blob/master/appendix.Rmd) file made up of groupings of a heading, some text, and sometimes a code chunk that points to a file, like a [template](https://github.com/ropensci-org/blog-guidance/tree/master/templates) or [checklist](https://github.com/ropensci-org/blog-guidance/tree/master/checklists) that populates that appendix.

<!--html_preserve--> {{< figure src = "file-comp.png" width = "800" alt = "this is the alternative text" >}}<!--/html_preserve-->

I often felt quite euphoric about the things I was able to figure out comparing GitHub file structures to their books. 

> (Consisentely, within days I would take this hard-earned knowledge for granted and feel inadequate in the face of my next technical challenge...until I felt the satisfaction of owning that next one too.)

Author: ```suggestion
Consistently, within days I would take this hard-earned knowledge for granted and feel inadequate in the face of my next technical challenge...until I felt the satisfaction of owning that next one too.
```

These bookdown books were helpful in exploring those comparisons:

> - rOpenSci Packages: Development, Maintenance, and Peer Review [GitHub](https://github.com/ropensci/dev_guide) _cf_ [Book](https://devguide.ropensci.org/)

Author: ```suggestion
- rOpenSci Packages: Development, Maintenance, and Peer Review [source on GitHub](https://github.com/ropensci/dev_guide) _cf_ [deployed book](https://devguide.ropensci.org/)
```

> - R for Data Science [GitHub](https://github.com/hadley/r4ds) _cf_ [Book](https://r4ds.had.co.nz/)

Author: ```suggestion
- R for Data Science [source on GitHub](https://github.com/hadley/r4ds) _cf_ [deployed book](https://r4ds.had.co.nz/)
```

> #### 5. How to everything

Author: Lol I want the title to remain this

Author: But a common theme between trouble shooting and linking issue is finding an error and fixing it. :thinking:

Reviewer: I like that title - was temporary but intentional. Cause it felt like this bookdown project really was learning "how to everything" 😆

(not sure what this tip should look like - it's a placeholder for things that don't fit)
What doesn't appear in this list of tips are all the things I've already forgotten that I had trouble with and learned to overcome by reading error messages and searching and poking at my setup until things worked. 
Looking back in Slack conversations with Maëlle I see that setting up to serve the bookdown preview using the RStudio bookdown Addin was tricky.
But errors usually had more to do with making sure the packages I was using were up to date and needing to install the development version of usethis. 

> I learned that adding "Fix #54" in a pull request description automatically closes issue #54 when the pull request is merged into master! (link to explanation?).

Author: ```suggestion
I learned that [adding "Fix #54" in a pull request description automatically closes issue #54 when the pull request is merged into master](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue)!.
```

```

And some of the further steps are skipped based on `fork`.

```yaml
      - name: Install Netlify CLI
        if: env.fork == 'false'
        run: npm install netlify-cli -g
```

If you're feeling motivated to add GitHub Actions deployment to your bookdown project, on top of Jim's video mentioned earlier and our workflows, be sure to note that Emil Hvitfeld wrote an [excellent guide to deploying your book on Netlify with GitHub Actions](https://www.hvitfeldt.me/blog/bookdown-netlify-github-actions/), with screenshots!

### Conclusion

In this post, we shared useful bookdown things we learnt whilst working on rOpenSci blog guidance. 
We're quite happy with both the end product and our learnings! 
If _you_ want to get started with bookdown, we'd recommend [the bookdown book](https://bookdown.org/yihui/bookdown/introduction.html), [this introductory slidedeck by Alison Hill](https://alison.rbind.io/talk/2019-rsc-bookdown/) and [the bookdown gallery](https://bookdown.org/home/archive/).

[^1]: The first bookdown book Maëlle worked on is [rOpenSci dev guide](https://devguide.ropensci.org/)!

> [^2]: Hat-tip to Julia Stewart Lowndes for using this phrasing to describe herself as if she read my mind.

Reviewer: I generally link to a person's rOpenSci author page if they have one. And I make sure it links to their website.

> [^2]: Hat-tip to Julia Stewart Lowndes for using this phrasing to describe herself as if she read my mind.

Author: Oh right :woman_facepalming:

GitHub file structure and contents of completed books with their corresponding live pages. This, for me, had the biggest payoff in learning how to bookdown.

Things I learned:

- One book chapter is made from one Rmd file.

- Chapters are woven together in the [`_bookdown.yml`](https://github.com/ropensci-org/blog-guidance/blob/master/_bookdown.yml) file that references those Rmd files

- Create links inside the book with `{#anchor}`. In the [`authorcontent.Rmd`](https://github.com/ropensci-org/blog-guidance/blob/master/authorcontent.Rmd) file, the heading `# Content Guidelines {#content}` means that a markdown format `[link to that chapter](#content)` anywhere in the book will link to the Content Guidelines chapter.

- Appendices A to H (_too many; we know_) are created from a single [appendix.Rmd](https://github.com/ropensci-org/blog-guidance/blob/master/appendix.Rmd) file made up of groupings of a heading, some text, and sometimes a code chunk that points to a file, like a [template](https://github.com/ropensci-org/blog-guidance/tree/master/templates) or [checklist](https://github.com/ropensci-org/blog-guidance/tree/master/checklists) that populates that appendix.

> <!--html_preserve--> {{< figure src = "file-comp.png" width = "800" alt = "this is the alternative text" >}}<!--/html_preserve-->

Reviewer: Image is blurry. Can you suggest a better way? Having one on top of other, instead of side by side prob takes up too much space.

Need to add alt text once decided

Reviewer: would an image like that be good for twitterImg? would have to have right proportions

Author: I don't find it blurry? I think it's fine?

#### 4. How to have alternative text but no captions for figures

In R Markdown, the same chunk options `fig.cap` controls the caption and alternative text of images.
We wanted [alternative text](https://www.w3.org/WAI/tutorials/images/informative/) but no caption.
The header option `figure_caption` didn't work.
I opened [an issue in bookdown repo](https://github.com/rstudio/bookdown/issues/856) after not getting solutions [on an RStudio community post](https://community.rstudio.com/t/have-an-alternative-text-but-no-caption-or-a-different-caption-in-bookdown-git-book/50370).

So if bookdown doesn't really support having alternative text but no captions for figures, what did we do? Thanks to a good tip by [Romain Lesur](https://github.com/RLesur), I wrote CSS to remove all captions.

```css
.caption {
  display: none;
}
```

The lines above live with other styling stuff in a file called `style.css`, that we refer to in [the `_output.yml` config file](https://github.com/ropensci-org/blog-guidance/blob/cae5887f392b4137eec50467779e9bba973b0b69/_output.yml#L3).

> #### 5. How to deploy a preview of the book for PRs

Author: I'll actually use "pull requests" throughout.

```{r masterworkflow, echo = FALSE}
details::details(readLines("https://raw.githubusercontent.com/ropensci-org/blog-guidance/master/.github/workflows/master.yml"), summary = "master workflow", lang = "yaml")
```

* after each commit in a PR from a fork, the book is built so we'd notice something breaking the Rmd. [log example](https://github.com/ropensci-org/blog-guidance/runs/466995315?check_suite_focus=true);

* after each commit in a PR from the repo, the book is built and deployed to a Netlify preview whose URL is posted in the PR checks. [Log example](https://github.com/ropensci-org/blog-guidance/runs/471968808?check_suite_focus=true), [Direct link to the check with the preview link](https://github.com/ropensci-org/blog-guidance/runs/471974005?check_suite_focus=true).

```{r prworkflow, echo = FALSE}
details::details(readLines("https://raw.githubusercontent.com/ropensci-org/blog-guidance/master/.github/workflows/pr.yml"), summary = "PR workflow", lang = "yaml")
```

Highlights from the PR workflow above

> * To deploy to Netlify _and_ get the preview URL, the workflow doesn't use Netlify's GitHub Actions but instead installs Netlify CLI, extracts the URL using `jq`[^4] and set it as an environment variable that can be used by next steps. I got the idea from [a thread on Netlify forum](https://community.netlify.com/t/deploying-preview-web-hook-via-api/3320/2).

Author: done locally

* To find out whether the PR is from a fork (and to skip all deploy steps for that, since forks don't have access to the Netlify secrets saved in our repo settings), I use `jq` on the raw info about the build, idea I got [from Vanessa Sochat](https://github.com/r-universe/vsoch/blob/2bf10b1e30f59f5fab64ec2b7332526c47f1f4d3/.github/workflows/pull-request-update-packages.yml#L26).

```yaml
      - name: Is this a fork
        run: |
          fork=$(jq --raw-output .pull_request.head.repo.fork "${GITHUB_EVENT_PATH}");echo "::set-env name=fork::$fork"
```

And some of the further steps are skipped based on `fork`.

```yaml
      - name: Install Netlify CLI
        if: env.fork == 'false'
        run: npm install netlify-cli -g
```

If you're feeling motivated to add GitHub Actions deployment to your bookdown project, on top of Jim's video mentioned earlier and our workflows, be sure to note that Emil Hvitfeld wrote an [excellent guide to deploying your book on Netlify with GitHub Actions](https://www.hvitfeldt.me/blog/bookdown-netlify-github-actions/), with screenshots!

### Conclusion

> In this post, we shared useful bookdown things we learnt whilst working on rOpenSci blog guidance. 

Reviewer: ```suggestion
In this post, we shared useful things we learned about bookdown and R project management  whilst working on the new [rOpenSci Blog Guidelines for Authors and Editors](https://blogguide.ropensci.org/). 
```

> We're quite happy with both the end product and our learnings! 

Reviewer: ```suggestion
We're almost as happy with our new skills as we are with the Blog Guide itself! 
```
